File: lib/morph/limit_output.rb
#!/usr/bin/env ruby

require "optparse"
require "open3"

# NOTE: This script MUST be compatible with ruby 1.9 through to current versions - do NOT modernize it!

module Morph

  # This wrapper script runs a command and lets standard out and error flow
  # through. However, it does limit the number of lines of output. Morph
  # uses this as a wrapper around running scrapers to ensure that they
  # can't fill up the docker container log file (and hence the server disk)

  class LimitOutput
    USAGE = "Usage: ./limit_output.rb max_lines command_to_run"

    def self.run(args, options = {})
      stdout_io = options.fetch(:stdout_io, $stdout)
      stderr_io = options.fetch(:stderr_io, $stderr)
      args = OptionParser.new do |opts|
        opts.banner = USAGE
        if args.length.zero?
          stderr_io.puts "Please give me the maximum number of lines"
        elsif args.length == 1
          stderr_io.puts "Please give me a command to run"
        elsif args.length > 2
          stderr_io.puts "Please give me the command to run as one argument"
        end
        if args.length < 2
          stderr_io.puts opts
          return 1
        end
      end.parse(args)

      max_lines = args[0].to_i
      command = args[1]

      # Disable output buffering
      stdout_io.sync = true
      stderr_io.sync = true

      stdout_buffer = ""
      stderr_buffer = ""

      line_count = 0

      exit_status = 1
      Open3.popen3(command) do |_stdin_pipe, stdout_pipe, stderr_pipe, wait_thr|
        streams = [stdout_pipe, stderr_pipe]
        until streams.empty?
          IO.select(streams).flatten.compact.each do |io|
            if io.eof?
              streams.delete io
              next
            end

            on_stdout_stream = io.fileno == stdout_pipe.fileno
            # Just send this stuff straight through
            buffer = on_stdout_stream ? stdout_buffer : stderr_buffer
            s = io.readpartial(1)
            buffer << s
            next unless s == "\n"

            if line_count < max_lines || max_lines.zero?
              (on_stdout_stream ? stdout_io : stderr_io) << buffer
            elsif line_count == max_lines
              stderr_io.puts "limit_output.rb: Too many lines of output!"
            end
            if on_stdout_stream
              stdout_buffer = ""
            else
              stderr_buffer = ""
            end
            line_count += 1
          end
        end

        # Output whatever is left in the buffers
        if line_count < max_lines || max_lines.zero?
          stdout_io << stdout_buffer
          stderr_io << stderr_buffer
        end

        process_status = wait_thr.value
        exit_status = process_status.exited? ? process_status.exitstatus : process_status.termsig + 128
      end

      exit_status
    end
  end
end

exit Morph::LimitOutput.run(ARGV) if __FILE__ == $0

File: lib/morph/limit_output.py
#!/usr/bin/env python3

import sys
import select
import subprocess
import argparse

# This wrapper script runs a command and lets standard out and error flow
# through. However, it does limit the number of lines of output. Morph
# uses this as a wrapper around running scrapers to ensure that they
# can't fill up the docker container log file (and hence the server disk)

def run(args, stdout_io=None, stderr_io=None):
    stdout_io = stdout_io or sys.stdout
    stderr_io = stderr_io or sys.stderr

    parser = argparse.ArgumentParser(
        usage="%(prog)s max_lines command_to_run",
        add_help=False
    )
    parser.add_argument('max_lines', type=int, nargs='?')
    parser.add_argument('command', nargs='?')

    try:
        parsed = parser.parse_args(args)
    except SystemExit:
        stderr_io.write("Usage: limit_output.py max_lines command_to_run\n")
        return 1

    if parsed.max_lines is None:
        stderr_io.write("Please give me the maximum number of lines\n")
        stderr_io.write("Usage: limit_output.py max_lines command_to_run\n")
        return 1
    if parsed.command is None:
        stderr_io.write("Please give me a command to run\n")
        stderr_io.write("Usage: limit_output.py max_lines command_to_run\n")
        return 1

    max_lines = parsed.max_lines
    command = parsed.command

    # Disable output buffering
    stdout_io.reconfigure(line_buffering=True) if hasattr(stdout_io, 'reconfigure') else None
    stderr_io.reconfigure(line_buffering=True) if hasattr(stderr_io, 'reconfigure') else None

    stdout_buffer = b""
    stderr_buffer = b""
    line_count = 0

    proc = subprocess.Popen(
        command,
        shell=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE
    )

    streams = {
        proc.stdout.fileno(): (proc.stdout, stdout_io, True),
        proc.stderr.fileno(): (proc.stderr, stderr_io, False)
    }

    while streams:
        readable, _, _ = select.select(list(streams.keys()), [], [])

        for fd in readable:
            pipe, output_io, is_stdout = streams[fd]

            try:
                chunk = pipe.read(1)
            except (OSError, ValueError):
                del streams[fd]
                continue

            if not chunk:
                del streams[fd]
                continue

            buffer = stdout_buffer if is_stdout else stderr_buffer
            buffer += chunk

            if chunk == b"\n":
                if line_count < max_lines or max_lines == 0:
                    output_io.buffer.write(buffer)
                    output_io.flush()
                elif line_count == max_lines:
                    stderr_io.write("limit_output.py: Too many lines of output!\n")
                    stderr_io.flush()

                if is_stdout:
                    stdout_buffer = b""
                else:
                    stderr_buffer = b""
                line_count += 1
            else:
                if is_stdout:
                    stdout_buffer = buffer
                else:
                    stderr_buffer = buffer

    # Output whatever is left in the buffers
    if line_count < max_lines or max_lines == 0:
        if stdout_buffer:
            stdout_io.buffer.write(stdout_buffer)
            stdout_io.flush()
        if stderr_buffer:
            stderr_io.buffer.write(stderr_buffer)
            stderr_io.flush()

    return_code = proc.wait()
    return return_code if return_code >= 0 else 128 + abs(return_code)

if __name__ == "__main__":
    sys.exit(run(sys.argv[1:]))
